### DISCUSSION TITLE
Running Java-based formulae with non-Homebrew JDK

### DISCUSSION THREAD

**Discussion by md5** [Homebrew/discussions#910] [Everyday usage] [answered]
It looks like the recent deprecation and almost immediate removal of `depends_on :java` and a bunch of subsequent formula updates have forced the install of `openjdk` for pretty much all Java-based formulas.

I was directed here by @jonchang and after doing some digging, found that this was clearly intentional:

* https://github.com/Homebrew/brew/pull/9403
* https://github.com/Homebrew/homebrew-core/issues/63290
* https://github.com/Homebrew/brew/pull/9052

My question is this: since it doesn't appear to be straightforward to avoid installing `openjdk` (though I made it work locally by changing to `depends_on "openjdk" => :recommended` in the formula I wanted to install, which `brew audit --strict` complained about), what is the recommended way to run Java-based utilities with a non-Homebrew JDK?

**Comment by SMillerDev** [comment]
There is no recommended way because the macOS system doesn't come with a JDK. You can install a JDK into the system but you'd never know what the version was.

The recommended way to depend on a JDK is to depend on the homebrew Formula. If that doesn't work for you, you could write your own Requirement class in the formula.

**Reply by jonchang** [reply]
See https://docs.brew.sh/Building-Against-Non-Homebrew-Dependencies and https://rubydoc.brew.sh/Requirement.html. We have a few custom Requirements in homebrew/core that you can take a peek at, or you can just copy the java requirement from git history into your own tap/formula.

**Reply by md5** [reply]
To be clear, I'm interested in installing an existing formula, `gradle`, without installing `openjdk`, not making my own formula. I could use my own forked formula, but then any other formula that requires `gradle` would still try to install `openjdk` instead of using my fork/tap of it.

**Reply by carlocab** [reply]
You would need to maintain your own version of those formulae too, then. This isn't very convenient, but it's not meant to be. (Sorry.) Switching out dependencies in homebrew/core isn't supported because it's not tested in CI.

You could also try passing `--ignore-dependencies` when installing `gradle` (and possibly its dependents too), but that setup is fragile. I'm not sure if an unqualified `brew upgrade`, for example, will respect that that flag was passed at the time you first installed a formula. (It might also do other weird things.)

**Reply by md5** [reply]
Got it. I understand that Homebrew has the policies it has to keep the amount of work for maintainers reasonable, so I'm not trying to ask for special treatment. That being said, it really sucks that as soon as anyone has any special requirement with a `homebrew-core` dependency they are immediately forced to either abandon Homebrew for that software or fork the formula and all of its dependents.

Looking at the formulas, the use of `Java.overridable_java_home_env` indicates that Homebrew expects users to set `JAVA_HOME` explicitly in order to use their own JDK for these formulas. Is that accurate? If so, is it documented anywhere? I guess people are supposed to set this in their shell profile or in the individual command invocation or something?

Also, could link to this discussion from https://github.com/Homebrew/brew/pull/10397? It's locked, so I can't comment any more. In particular, Mike mentioned that this removal was in the release notes, but I did not find it mentioned in either of these places:

* https://brew.sh/2021/02/05/homebrew-3.0.0/
* https://github.com/Homebrew/brew/releases/tag/3.0.0

BTW, just signed up as a sponsor. Keep up the good work! 👍

**Reply by carlocab** [reply]
Yes, if a formula has `overridable_java_home_env`, then if you set your `JAVA_HOME` manually, the software installed by that formula should use whichever Java your `JAVA_HOME` points to. You can set it in whatever manner is convenient for you.

Not sure about the documentation for this. If it's not there, or inadequately documented, PRs to improve them would be appreciated.

Thanks for sponsoring Homebrew!

**Reply by jonchang** [reply]
It's not fair to characterize the deprecation and removal of Java requirements as being immediate, sudden, or without warning. We started migrating formulae in late October and finished in late November, at which point Homebrew would have started warning you about the impending deprecation and removal of this feature.

The relevant announcements are in the 2.5.9 release notes and highlighted in the 2.6.0 blog post. There was another reminder in the 2.7.0 post and this function was disabled in 3.0.0.

Homebrew is a rolling release package manager and things will change quickly. Four months of notice is actually an unusually long time for us to turn off something like this. If that's not sufficient time then consider using a different package manager with stable point releases of its package index.

**Reply by md5** [reply]
> The relevant announcements are in the 2.5.9 release notes and highlighted in the 2.6.0 blog post. There was another reminder in the 2.7.0 post and this function was disabled in 3.0.0.

Given that Homebrew implicitly updates major versions, I'm not sure how users are expected to see that, but thanks for pointing out which release notes contained the deprecation notice. It seems like the removals themselves should be in the 3.0.0 release notes, though.

> Homebrew is a rolling release package manager and things will change quickly. Four months of notice is actually an unusually long time for us to turn off something like this. If that's not sufficient time then consider using a different package manager with stable point releases of its package index.

Understood, but your response seems disingenuous given that there aren't any other Macos package managers with anything close to Homebrew's adoption or coverage. Still, I apologize for any lack of consideration in my original post and greatly appreciate the work you all do, as I mentioned before.

**Reply by md5** [reply]
> Yes, if a formula has `overridable_java_home_env`, then if you set your `JAVA_HOME` manually, the software installed by that formula should use whichever Java your `JAVA_HOME` points to. You can set it in whatever manner is convenient for you.
> 
> Not sure about the documentation for this. If it's not there, or inadequately documented, PRs to improve them would be appreciated.

Where would be the right place to document that if I wanted to create a PR?

**Reply by carlocab** [reply]
>there aren't any other Macos package managers with anything close to Homebrew's adoption or coverage

I think a big part of why this is true is that Homebrew very aggressively targets a narrow set of use cases that applies to as large a group of users are possible. This leads to things like not being able to swap out dependencies easily. So there might be a catch-22 type situation here.

>Where would be the right place to document that if I wanted to create a PR?

The documentation you see at https://docs.brew.sh is hosted here: https://github.com/Homebrew/brew/tree/master/docs

**Reply by md5** [reply]
It seems like the `caveats` section of each formula using `openjdk` is one way to do this consistently. I could imagine having a `#java_home_caveats` method on `Language::Java` that could be used to provide boilerplate caveats for the formulae. Are there any other groups of formulae with standard, boilerplate `caveats` or is there another mechanism like `caveats` that would be preferable?

**Reply by carlocab** [reply]
I don't think `caveats` are a good place for this. This doesn't come up very often (admittedly, it might be because a lot of users don't know about it...), and, as it is, lots of people ignore the messages that `brew` prints out. (Anecdotally, I would say about 80% of support requests I read are from people who didn't read error/warning/instructional messages properly.) Printing more messages that most people ignore is likely to just make that worse.

**Reply by md5** [reply]
> I don't think `caveats` are a good place for this. This doesn't come up very often (admittedly, it might be because a lot of users don't know about it...), and, as it is, lots of people ignore the messages that `brew` prints out. (Anecdotally, I would say about 80% of support requests I read are from people who didn't read error/warning/instructional messages properly.) Printing more messages that most people ignore is likely to just make that worse.

That makes sense. If this were to go into `docs`, would you envision having a separate page for Java-based formulas overall that mentions this, perhaps along with advice to export `JAVA_HOME=$(/usr/libexec/java_home)` in order to use a self-installed JDK? I didn't see any other pages in the docs that look like they'd be a fit for adding this information.

Maybe FAQ or Common Issues, though I'm not sure this is either. Or maybe a page like this: https://docs.brew.sh/Homebrew-and-Python

**Reply by carlocab** [reply]
I'm not sure, and I don't have a strong opinion about it, since I don't really use Java. I think your perspective as a user might be more useful here.

Switching out dependencies has come up a few times, so perhaps the documentation regarding that can be expanded, and it could have a small section on using a different JDK if you're not too fussy about having multiple JDKs installed (i.e. by setting `JAVA_HOME` appropriately).

**Reply by danielnachun** [reply]
Out of curiosity, why are you trying not to use brewed OpenJDK?  We do have versioned support for some older ones so I'm not sure what other reasons there are for installing your own JDK.

**Reply by md5** [reply]
> Out of curiosity, why are you trying not to use brewed OpenJDK? We do have versioned support for some older ones so I'm not sure what other reasons there are for installing your own JDK.

I'm interested in using Java Flight Recorder, so I already had Oracle's Java 15 installed. That being said, it looks like OpenJDK may have gained parity there in the past year. I'm just refreshing my Java toolchain after not having touched it in a few years.

I'm not opposed across the board to using OpenJDK or something, but I originally wanted to understand why it was being installed and ended up digging further into the history of this change than I expected.

**Reply by danielnachun** [reply, marked as answer]
Apparently OpenJDK and Oracle JDK should now be interchangeable: https://dzone.com/articles/using-java-flight-recorder-with-openjdk-11-1.  So you may be able to use brewed OpenJDK to do everything you want!

**Reply by md5** [reply]
> I'm not sure, and I don't have a strong opinion about it, since I don't really use Java. I think your perspective as a user might be more useful here.
> 
> Switching out dependencies has come up a few times, so perhaps the documentation regarding that can be expanded, and it could have a small section on using a different JDK if you're not too fussy about having multiple JDKs installed (i.e. by setting `JAVA_HOME` appropriately).

I've opened a PR at https://github.com/Homebrew/brew/pull/10700. We can iterate there or if you all don't want a "Homebrew and Java" page I can close it 👍
